### PR TITLE
gpperfmon: Fix incorrect data saved into queries_history table

### PR DESCRIPTION
**Pull request by Chibin**
sigar_proc_cpu_get is the function which takes a pid
    and goes to the cpu and gets the cpu_elapsed value for that pid at a
    given moment.

    get_pid_metrics saves the information into a hash table within the
    gpsmon process. That pid only gets wiped out of the hashtable when
    gpmmon requests a dump command to gpsmon.

    gpsmon process sends tcp packets with information about cpu_elapsed
    (amount of cpu cycles spent on a given
    slice/process). Doing a 'dump' clears out the gpsmon hashtables.

    This could lead to scenarios where a query has ended,
    the process has died, and the sigar_proc_cpu_get function can't find the pid
    so it puts 0 into the struct which is updating the hashtable so the
    hashtable has a 0 for cpu_elapsed and then this is what is in
    queries_tail and so queries_history ends up having a 0 for cpu_elapsed
    since it was the last entry in queries_tail

    Fix: Ensure that we validate the functions that request pid metrics
    from libsigar and log the issues if they occur.

### CODE REVIEW

File: gpAux/gpperfmon/src/gpmon/gpsmon.c [hunk 1]
```diff
@@ -259,13 +260,29 @@ static void get_pid_metrics(apr_int32_t pid, apr_int32_t tmid, apr_int32_t ssid,
...
+		return;
+	}
+
+	status = sigar_proc_fd_get(gx.sigar, pid, &fd);
+	if (status != SIGAR_OK) {
+		TR0(("[WARNING] %s. PID: %d\n", strerror(status), pid));
+		return;
+	}
+
+	TR0(("--------------------- starting %d\n", pid));
```
Comment (danielgustafsson) [inline gpAux/gpperfmon/src/gpmon/gpsmon.c]: What is the difference between `TR0()` and `TR2()` which was previously used? (I assume the wording change is fixing a typo.)

Comment (Chibin) [inline gpAux/gpperfmon/src/gpmon/gpsmon.c]: > What is the difference between TR0() and TR2() which was previously used? (I assume the wording change is fixing a typo.)

Good catch! Yes, this is definitely a typo. TR0 is our macro to print to our gpmmon log, INFO level.
TR2 is more for DEBUG level logging.

File: gpAux/gpperfmon/src/gpmon/gpsmon.c [hunk 2]
```diff
@@ -259,13 +260,29 @@ static void get_pid_metrics(apr_int32_t pid, apr_int32_t tmid, apr_int32_t ssid,
...
 	memset(&cpu, 0, sizeof(cpu));
 	memset(&mem, 0, sizeof(mem));
 	memset(&fd, 0, sizeof(fd));
-	sigar_proc_mem_get(gx.sigar, pid, &mem);
-	sigar_proc_cpu_get(gx.sigar, pid, &cpu);
-	sigar_proc_fd_get(gx.sigar, pid, &fd);
+
+	status = sigar_proc_mem_get(gx.sigar, pid, &mem);
+	if (status != SIGAR_OK) {
+		TR0(("[WARNING] %s. PID: %d\n", strerror(status), pid));
```
Comment (larham) [inline gpAux/gpperfmon/src/gpmon/gpsmon.c]: I'm wondering about printing a warning here.  If the situation is expected/normal, like a query that is finished but not "dumped" yet (so no process exists), then it doesn't feel like a warning.  Perhaps conditionalize the cases when the result of sigar is truly unexpected?

File: gpMgmt/test/behave/mgmt_utils/gpperfmon.feature [hunk 3]
```diff
@@ -149,25 +137,16 @@ Feature: gpperfmon
         When below sql is executed in "gptest" db
         """
         CREATE TABLE sales (id int, date date, amt decimal(10,2))
-        DISTRIBUTED BY (id)
-        PARTITION BY RANGE (date)
-        ( START (date '2016-11-01') INCLUSIVE
-          END (date '2016-12-02') EXCLUSIVE
-          EVERY (INTERVAL '1 day') );
+        DISTRIBUTED BY (id);
```
Comment (larham) [inline gpMgmt/test/behave/mgmt_utils/gpperfmon.feature]: I like how this simplifies the test case.

**Comment by Chibin** [comment]
Looks like there's some flake in the gpperfmon behave test.
Looking into the issue before being able to merge this in.

**Comment by Chibin** [comment]
Thanks for the review!

**Pull request merged**